src/parser/mod.rs

```diff
@@ -11679,14 +11679,21 @@ impl<'a> Parser<'a> {
     pub fn parse_update(&mut self) -> Result<Statement, ParserError> {
         let or = self.parse_conflict_clause();
         let table = self.parse_table_and_joins()?;
+        let from_before_set = if self.parse_keyword(Keyword::FROM)
+            && dialect_of!(self is GenericDialect | PostgreSqlDialect | DuckDbDialect | BigQueryDialect | SnowflakeDialect | RedshiftSqlDialect | MsSqlDialect | SQLiteDialect )
+        {
+            Some(self.parse_table_and_joins()?)
+        } else {
+            None
+        };
         self.expect_keyword(Keyword::SET)?;
         let assignments = self.parse_comma_separated(Parser::parse_assignment)?;
         let from = if self.parse_keyword(Keyword::FROM)
             && dialect_of!(self is GenericDialect | PostgreSqlDialect | DuckDbDialect | BigQueryDialect | SnowflakeDialect | RedshiftSqlDialect | MsSqlDialect | SQLiteDialect )
         {
             Some(self.parse_table_and_joins()?)
         } else {
-            None
+            from_before_set
```

Comment:
Could we instead introduce a wrapper type to encode the position information?
Something like
```rust
enum UpdateTableFromKind {
     BeforeSet(TableWithJoins),
     AfterSet(TableWithJoins)
}
...
Statement::Update {... from: Option<UpdateTableFromKind> }
```
See [TableSampleKind](https://github.com/yoavcloud/datafusion-sqlparser-rs/blob/eae5629fb86f5e262063c0bc99ff628a5855168f/src/ast/query.rs#L1157-L1162) for an example on this approach.

That would ensure that we can produce the same sql as in the input.

Reply:
@iffyio My problem with your approach that it will break backwards compatibility, and I don't feel it "worth it" in this case. What do you think?

Reply:
Ah yeah I think that would be ok in this case and the change required downstream would be mechanical

Reply:
@iffyio I will change that but I have to say that I disagree. I have a system on top of sqlparser and every kind of this braking change is causing bugs and takes time to migrate.

Reply:
@alamb fyi